src/sentry/utils/appleconnect/itunes_connect.py

```diff
@@ -349,3 +320,396 @@ def get_dsym_url(
+        """
+        client = cls(service_key)
+        client.load_session_cookie(cookie)
+        if client.state is not ClientState.AUTHENTICATED:
+            raise SessionExpiredError
+        return client
+
+    def request_auth_service_key(self) -> ITunesServiceKey:
+        """Obtains the authentication service key used in X-Apple-Widget-Key header.
+
+        This key is pretty static and can be cached long-term, saving a request to fetch it.
+        It can optionally be provided to :meth:`init`.
+
+        :returns: the service key to be used in all future calls X-Apple-Widget-Key headers.
+
+        :raises: any exception in case of failure.
+        """
+        url = "https://appstoreconnect.apple.com/olympus/v1/app/config?hostname=itunesconnect.apple.com"
+        logger.debug("GET %s", url)
+        response = self.session.get(url)
+        data = response.json()
+        return ITunesServiceKey(data["authServiceKey"])
+
+    def start_login_sequence(self, username: str, password: str) -> None:
+        """Starts the login process.
+
+        This will trigger the two factor authentication code to show on any validated device
+        for the user.  The code must be provided using :meth:`auth_code`.
+
+        Alternatively after calling this you can switch to SMS two factor authentication by
+        calling :meth:`request_sms`.
+        """
+        assert self.state in [ClientState.NEW, ClientState.EXPIRED]
+        url = "https://idmsa.apple.com/appleauth/auth/signin"
+        logger.debug("POST %s", url)
+
+        start_login = self.session.post(
+            url,
+            json={
+                "accountName": username,
+                "password": password,
+                "rememberMe": True,
+            },
+            headers={
+                "X-Requested-With": "XMLHttpRequest",
+                "X-Apple-Widget-Key": self.service_key,
+                "Accept": "application/json, text/javascript",
+            },
+        )
+
+        if start_login.status_code == http.HTTPStatus.OK:
+            raise ITunesError(
+                f"Two factor auth not enabled for user, status_code={start_login.status_code}"
+            )
+        if start_login.status_code == http.HTTPStatus.CONFLICT:
+            self._session_id = start_login.headers["x-apple-id-session-id"]
+            self._scnt = start_login.headers["scnt"]
+            self.state = ClientState.AUTH_REQUESTED
+        else:
+            raise ITunesError(f"Unexpected status code form signing: {start_login.status_code}")
+
+    def two_factor_code(self, code: str) -> None:
+        """Sends the two-factor authentication code, completing authentication.
+
+        :raises: :class:`InvalidTwoFactorAuthError`
+        """
+        assert self.state is ClientState.AUTH_REQUESTED
+        url = "https://idmsa.apple.com/appleauth/auth/verify/trusteddevice/securitycode"
+        logger.debug(f"POST {url}")
```

Comment:
please keep this consistent with the rest of such logs. tbh, I don’t know which to prefer here, f-strings, or `%s` interpolation?

Reply:
ah, good catch.  this is mostly copied form the existing implementations and i've been trying to fix this.  In python you should still use % formatting for logging calls, because it defers the formatting until, and if, the log is actually emitted.  So when prod runs on INFO level it saves some CPU cycles to not do the formatting.